oauthenticator/github.py

```diff
@@ -88,18 +119,40 @@ def authenticate(self, handler, data=None):
         access_token = resp_json['access_token']
         
         # Determine who the logged in user is
-        headers={"Accept": "application/json",
-                 "User-Agent": "JupyterHub",
-                 "Authorization": "token {}".format(access_token)
-        }
         req = HTTPRequest("https://%s" % GITHUB_API,
                           method="GET",
-                          headers=headers
+                          headers=_api_headers(access_token)
                           )
         resp = yield http_client.fetch(req)
         resp_json = json.loads(resp.body.decode('utf8', 'replace'))
 
-        return resp_json["login"]
+        username =  resp_json["login"]
+
+        # Check if user is a member of any whitelisted organizations.
+        # This check is performed here, as it requires `access_token`.
+        if self.github_organization_whitelist:
+            user_in_org = yield self._check_organization_whitelist(username, access_token)
+            return username if user_in_org else None
+        else:  # no organization whitelisting
+            return username
+
+
+    @gen.coroutine
+    def _check_organization_whitelist(self, username, access_token):
+        http_client = AsyncHTTPClient()
+        headers = _api_headers(access_token)
+        # Get all the user's organizations
+        next_page = "https://%s/orgs" % GITHUB_API
+        user_orgs = set()
+        while next_page:
```

Comment:
This can check the orgs on each page and return immediately if a match is found, rather than walking through all pages every time:

```python
while next_page:
    ...
    user_orgs = set(entry["login"] for entry in resp_json)
    if self.github_organization_whitelist & user_orgs:
        return True
return False
```